qmt/data/solvers_2d.py

```diff
@@ -2,43 +2,40 @@
 # Licensed under the MIT License.
 
 from collections import namedtuple
+from typing import NamedTuple, Optional, Tuple
+from sympy.core.mul import Mul
+import numpy as np
+from qmt.physics_constants import UArray
 
-Potential2dData = namedtuple(
-    "Potential2dData",
-    ["coordinates", "coordinate_units", "potential", "potential_units"],
-)
-ThomasFermi2dData = namedtuple(
-    "ThomasFermi2dData",
-    [
-        "coordinates",
-        "potential",
-        "density",
-        "conduction_band",
-        "valence_band",
-        "reference_level",
-        "temperature",
-        "coordinate_units",
-        "potential_units",
-        "energy_units",
-    ],
-)
 
-Bdg2dData = namedtuple(
-    "Bdg2dData",
-    [
-        "coordinates",
-        "energies",
-        "wave_functions",
-        "coordinate_units",
-        "energy_units",
-        "hamiltonian",
-    ],
-)
+class Potential2dData(NamedTuple):
+    coordinates: Tuple[UArray, UArray]
+    potential: UArray
 
 
-Phase2dData = namedtuple(
-    "Phase2dData", ["coordinates", "superconducting_phase", "coordinate_units"]
-)
+# Conceptually this is ThomasFermi2dData(Potential2dData), but python doesn't let us do
+# that yet :(
+class ThomasFermi2dData(NamedTuple):
+    coordinates: Tuple[UArray, UArray]
+    potential: UArray
+    density: UArray
+    conduction_band: UArray
+    valence_band: UArray
+    reference_level: Mul
+    temperature: Mul
+
+
+class Bdg2dData(NamedTuple):
+    coordinates: Tuple[UArray, UArray]
+    energies: UArray
+    wave_functions: UArray
+    hamiltonian: Optional[UArray]
+
+
+class Phase2dData(NamedTuple):
+    coordinates: Tuple[UArray, UArray]
```

Comment:
Afaik the reason why this is in QMT is so that external people (e.g. academic collaborators) who cannot access QMS can still make sense of the results of our simulations. For example: someone hands us a device design, we run the simulations for it using QMS, and then hand them the results.

By debug crap do you mean the raw Hamiltonian (that's the only one I've seen)? I suppose the technically correct way to do it is to only put publicly available information here in these classes, then make another class in QMS that inherits from these, and subclass those to include debug stuff. Although to make things simpler I think we might just want to include the debug stuff in the signatures here too and mark them as Optional (if that's not considered a security/IP leak), since python inheritance from NamedTuple doesn't seem to work yet...

Reply:
Well, sometimes, we, say, calculate density, or we need to integrate wavefunctions along some axis, etc. Right now, I basically return some dict with some values from bdg and some extra data.
But the dicts are mutable and they can't be cached so it's a problem. 

We can put extra as a frozenset of dict(), so that we can put whatever in it.

Reply:
Ok how about we use dataclasses? It's in python 3.7 but there's a backport available for 3.6
![Annotation 2019-04-02 143943](https://user-images.githubusercontent.com/4095626/55437995-4170bf80-5555-11e9-82fb-cbcbeeeabcf3.jpg)

Reply:
Yes, that's great!

Reply:
Changed to dataclasses now